### PR TITLE
KTLN-697: Difference Between Two Time Periods in Kotlin

### PR DESCRIPTION
**Pull request by tapankavasthi**
## Summary
Find Difference Between Two Time Periods in Kotlin

## Jira
[KTLN-697](https://jira.baeldung.com/browse/KTLN-697)

### CODE REVIEW

File: core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    fun `given two duration periods, when find difference using minus method, then give correct duration value`() {
+        val startDuration = 1.days + 12.hours + 15.minutes
+        val endDuration = 8.hours + 45.minutes
+        val difference: Duration = startDuration.minus(endDuration)
+        val days: Long = difference.inWholeDays
+        val hours: Long = difference.inWholeHours - days.days.inWholeHours
+        val mins: Long = difference.inWholeMinutes - days.days.inWholeMinutes - hours.hours.inWholeMinutes
+        assertEquals(1L, days)
+        assertEquals(3L, hours)
+        assertEquals(30L, mins)
```
Comment (theangrydev) [inline core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt:24]: think this assertion would be clearer on the duration itself, what do you think?
```suggestion
        assertEquals(1.days + 3.hours + 30.minutes, difference)
```

File: core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        val mark2 = timeSource.markNow()
+        val diff = mark2 - mark1
+        val seconds: Long = diff.inWholeSeconds
+        assertEquals(1L, seconds)
+    }
+
+    @Test
+    fun `given two epoch values, when find difference using Instant creation, then give correct duration value`() {
+        val startEpochMillis: Long = 1700617592000L
+        val endEpochMillis: Long = 1700617692000L
```
Comment (theangrydev) [inline core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt:41]: It's already clear the values are Long from the L suffix
```suggestion
        val startEpochMillis = 1700617592000L
        val endEpochMillis = 1700617692000L
```

File: core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        val instant1 = Instant.fromEpochMilliseconds(startEpochMillis)
+        val instant2 = Instant.fromEpochMilliseconds(endEpochMillis)
+        val durationDiff = instant2.minus(instant1)
+        val mins = durationDiff.inWholeMinutes
+        val secs = durationDiff.inWholeSeconds - mins.minutes.inWholeSeconds
+        assertEquals(1L, mins)
+        assertEquals(40L, secs)
```
Comment (theangrydev) [inline core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt:48]: I think this assertion would be clearer on the duration itself, what do you think?
```suggestion
        assertEquals(1.mins + 40.seconds, durationDiff)
```

**Review by theangrydev — CHANGES_REQUESTED** [review]

File: core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    @Test
+    fun `given two time marks, when find difference, then give correct time difference`() {
+        val timeSource = TimeSource.Monotonic
+        val mark1 = timeSource.markNow()
+        Thread.sleep(1000) // Sleep 0.5 seconds.
+        val mark2 = timeSource.markNow()
+        val diff = mark2 - mark1
+        val seconds: Long = diff.inWholeSeconds
+        assertEquals(1L, seconds)
+    }
```
[the remark this answers is not in the document]
Comment (tapankavasthi) [reply core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt]: I've addressed this concern while keeping the code simple to understand for the reader.

File: core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        assertEquals(1.days + 3.hours + 30.minutes, difference)
+    }
+
+    @Test
+    fun `given two epoch values, when find difference using Instant creation, then give correct duration value`() {
+        val startEpochMillis = 1700617592000L
+        val endEpochMillis = 1700617692000L
+        val instant1 = Instant.fromEpochMilliseconds(startEpochMillis)
+        val instant2 = Instant.fromEpochMilliseconds(endEpochMillis)
+        val durationDiff = instant2.minus(instant1)
```
Comment (theangrydev) [inline core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt:39]: For the sake of consistency, can we name this variable `difference` too

File: core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+class TimePeriodDifferenceUnitTest {
+
+    @Test
+    fun `given two time marks, when find difference, then give correct time difference`() {
+        val timeSource = TimeSource.Monotonic
+        val mark1 = timeSource.markNow()
+        val mark2 = mark1 + 5.seconds // use timeSource.markNow() to mark current time
+        val difference = mark2 - mark1
+        assertTrue(5.seconds <= difference)
+    }
```
Comment (theangrydev) [inline core-kotlin-modules/core-kotlin-date-time/src/test/kotlin/com/baeldung/timedifference/TimePeriodDifferenceUnitTest.kt:23]: Hmm. So you avoided the second `markNow()` to make the test predictable, but now it's a bit hard to understand what is being tested. How about something like this instead:
```suggestion
    @Test
    fun `given two time marks, when find difference, then give correct time difference`() {
        val timeSource = TimeSource.Monotonic
        val mark1 = timeSource.markNow()
        // call operation to measure duration of
        val mark2 = timeSource.markNow()
        val difference = mark2 - mark1
        assertThat(difference).isGreaterThanOrEqualTo(0);
    }
```

**Review by theangrydev — CHANGES_REQUESTED** [review]

**Review by theangrydev — APPROVED** [review]

**Pull request merged**